README.md
# **laura**: Local Auto-Regressive Average

This repository contains the source code for the **L**ocal
**AU**to-**R**egressive **A**verage (LAURA) inverse solution as described by de
Peralta Menendez et al. (2001, 2004). The code is based on
[mne-python](https://mne.tools/), a powerful EEG library for
[python](https://python.org/).

Personally, I think this linear inverse solution finds the neural sources
underlying M/EEG measurements with great success and is a valuable option among
other inverse solutions such as *minimum norm estimates* and *(e)LORETA*.

# Dependencies
* [mne](https://mne.tools/stable/index.html)
* [scipy](https://scipy.org/)
* [numpy](https://numpy.org/)  

That's it!
<br/>

# Installation from PyPi
Use [pip](https://pip.pypa.io/en/stable/) to install laura and all its
dependencies from [PyPi](https://pypi.org/project/laura/):

```
pip install laura
```

<br/>

# Quick Start
The following code demonstrates how to use this package:

```
from laura import compute_laura

stc = compute_laura(evoked, forward)
stc.plot()
```
, where evoked is an instance of **mne.Evoked** and forward is an instance of
**mne.Forward**. For further explanation on mne and its objects please refer to
the [mne website](https://mne.tools/).

For a more comprehensive tutorial hop over to [this notebook!](tutorials/tutorial_1.ipynb)

# Feedback
Please leave your feedback and bug reports at lukas_hecker@web.de.

<br/>

# References
Please cite the authors of the LAURA inverse solution appropriately:

[1] Menendez, R. G. D. P., Andino, S. G., Lantz, G., Michel, C. M., &
    Landis, T. (2001). Noninvasive localization of electromagnetic epileptic
    activity. I. Method descriptions and simulations. Brain topography, 14(2),
    131-137.

[2] de Peralta Menendez, R. G., Murray, M. M., Michel, C. M., Martuzzi, R.,
& Andino, S. L. G. (2004). Electrical neuroimaging based on biophysical
constraints. Neuroimage, 21(2), 527-539.

<br/>

I would be happy if you would cite this package, too:

```
LAURA was calculated using the laura python package available at https://github.com/LukeTheHecker/laura.
```

# Limitations
The current implementation is limited to:

* fixed dipole orientations
* time-domain EEG data
  
Feel free to modify the code and start a pull request!

# Troubleshooting
* Having problems with the installation? Check the [package requirements](requirements.txt).

changelog.md
# Changelog

## 0.0.2 30.05.2022
* split up the calculation of the inverse operator and the application to the data:  
  * make_laura -> calculates the inverse operator
  * apply_laura -> applies the inverse operator to the data
  * 
## 0.0.1 29.05.2022
* initial release

laura/tests/tests.py
import os
import mne
import numpy as np
from ..laura import compute_laura, make_laura, apply_laura

verbose = 0
spacing = 'ico2'
fs_dir = mne.datasets.fetch_fsaverage(verbose=verbose)
subjects_dir = os.path.dirname(fs_dir)

# The files live in:
subject = 'fsaverage'
trans = os.path.join(fs_dir, 'bem', 'fsaverage-trans.fif')
src = os.path.join(fs_dir, 'bem', 'fsaverage-ico-5-src.fif')
bem = os.path.join(fs_dir, 'bem', 'fsaverage-5120-5120-5120-bem-sol.fif')



# Create our own info object, see e.g.:

montage = mne.channels.make_standard_montage("standard_1020")
sfreq = 1000 
info = mne.create_info(montage.ch_names, sfreq, 
    ch_types=['eeg']*len(montage.ch_names), verbose=0)
info.set_montage("standard_1020")

# Create and save Source Model
src = mne.setup_source_space(subject, spacing=spacing, surface='white',
                                    subjects_dir=subjects_dir, add_dist=False,
                                    n_jobs=-1, verbose=verbose)

# Forward Model
fwd = mne.make_forward_solution(info, trans=trans, src=src,
                                bem=bem, eeg=True, mindist=5.0, n_jobs=-1,
                                verbose=verbose)
fwd = mne.convert_forward_solution(fwd, surf_ori=True, force_fixed=True,
                                    use_cps=True, verbose=verbose)
leadfield = fwd['sol']['data']

def test_laura_quick():
    data = np.random.randn(leadfield.shape[0], 10)
    evoked = mne.EvokedArray(data, info)
    stc = compute_laura(evoked, fwd)


def test_laura():
    data = np.random.randn(leadfield.shape[0], 10)
    evoked = mne.EvokedArray(data, info)
    G = make_laura(fwd)
    stc = apply_laura(evoked, G, fwd)


laura/laura.py
from scipy.spatial.distance import cdist
from scipy.linalg import inv
import numpy as np
import mne

def pos_from_forward(forward, verbose=0):
    ''' Get vertex/dipole positions from mne.Forward model

    Parameters
    ----------
    forward : instance of mne.Forward
        The forward model. 
    
    Return
    ------
    pos : numpy.ndarray
        A 2D matrix containing the MNI coordinates of the vertices/ dipoles

    Note
    ----
    forward must contain some subject id in forward["src"][0]["subject_his_id"]
    in order to work.
    '''
    # Get Subjects ID
    subject_his_id = forward["src"][0]["subject_his_id"]
    src = forward["src"]

    # Extract vertex positions from left and right source space
    pos_left = mne.vertex_to_mni(src[0]["vertno"], 0, subject_his_id, verbose=verbose)
    pos_right = mne.vertex_to_mni(src[1]["vertno"], 1, subject_his_id, verbose=verbose)

    # concatenate coordinates from both hemispheres
    pos = np.concatenate([pos_left, pos_right], axis=0)

    return pos

# def compute_laura(evoked, forward, noise_cov=None, reg=200, drop_off=2,
#     verbose=0):
#     """ Calculate Local AUto-Regressive Averages (LAURA) inverse solution


#     Parameters
#     ----------
#     evoked : instance of mne.Evoked
#         The evoked data
#     forward : instance of Forward
#         The forward solution.
#     noise_cov : instance of Covariance
#         The noise covariance.
#     reg : float
#         Value that weights noise regularization. The higher, the more
#         conservative the source localization will be. As a rule of thumb you can
#         estimate the optimal reg parameter as follows: reg = 1e3 / SNR, whereas
#         SNR corresponds to the signal-to-noise ratio in your data.
#     drop_off : int/float
#         Value that is often denoted as exponent "e_i" in the literature [2,
#         Appendix] and is set to "2" at default. It corresponds to fields that
#         decrease with the square of the inverse distance. Lowering this value
#         (e.g., drop_off=1.5) will decrease sparsity of sources, whereas
#         increasing this value (e.g., drop_off=2.5) will increase sparsity.
#     verbose : bool,
#         Controls verbosity.

#     Returns
#     -------
#     stc : instance of SourceEstimate
#         The source estimates.
    
#     References
#     ---------

#     [1] Menendez, R. G. D. P., Andino, S. G., Lantz, G., Michel, C. M., &
#     Landis, T. (2001). Noninvasive localization of electromagnetic epileptic
#     activity. I. Method descriptions and simulations. Brain topography, 14(2),
#     131-137.

#     [2] de Peralta Menendez, R. G., Murray, M. M., Michel, C. M., Martuzzi, R.,
#     & Andino, S. L. G. (2004). Electrical neuroimaging based on biophysical
#     constraints. Neuroimage, 21(2), 527-539.

    
#     """
#     assert forward['surf_ori'], "dipole orientations in mne.Forward must be fixed!"
#     evoked = evoked.copy().set_eeg_reference('average', proj=True).apply_proj()
#     forward = forward.copy()

#     # make sure to select all the right channels
#     if noise_cov is not None:
#         noise_cov = noise_cov.copy()
#         ch_names = list(set(noise_cov.ch_names).intersection(evoked.ch_names))
#         ch_names = list(set(ch_names).intersection(forward.ch_names))

#         noise_cov = noise_cov.pick_channels(ch_names)
#         evoked = evoked.pick_channels(ch_names)
#         forward = forward.pick_channels(ch_names)
#     else:
#         ch_names = list(set(forward.ch_names).intersection(evoked.ch_names))
#         evoked = evoked.pick_channels(ch_names)
#         forward = forward.pick_channels(ch_names)

#     # Extract the necessary variables from the forward model
#     pos = pos_from_forward(forward, verbose=verbose)
#     leadfield = forward['sol']['data']
#     vertices = [forward["src"][0]['vertno'], forward["src"][1]['vertno']]
  
#     # Select channels of interest
#     # sel = [evoked.ch_names.index(name) for name in ch_names]
#     x = evoked.data#[sel]
#     # ensure common average reference:
#     x -= x.mean(axis=0)

#     # pairwise distance matrix of all vertex/dipole locations
#     d = cdist(pos, pos)
#     # Get the adjacency matrix of the source spaces
#     adj = mne.spatial_src_adjacency(forward["src"], verbose=verbose).toarray()
#     # set non-neighboring dipoles to zero
#     for i in range(d.shape[0]):
#         # find dipoles that are no neighbor to dipole i
#         non_neighbors = np.where(~adj.astype(bool)[i, :])[0]
#         # append dipole itself
#         non_neighbors = np.append(non_neighbors, i)
#         # set non-neighbors to zero
#         d[i, non_neighbors] = 0
#         # neighbors = np.where(adj.astype(bool)[i, :])[0]
#         # d[i, neighbors] = -d[i, neighbors]**(drop_off)

#     A = -d**-drop_off
#     A[np.isinf(A)] = 0
#     W = np.identity(A.shape[0])
#     M = np.matmul(W, A)

#     # Source Space metric
#     W_j = np.matrix(inv(np.matmul(M.T, M)))
#     W_j_inv = np.matrix(inv(W_j))

#     # Data Space metric
#     if noise_cov is None:
#         W_d = np.matrix(np.identity(leadfield.shape[0]))
#     else:
#         # check if mne.Covariance is 1D (only diagonal values of the matrix)
#         if len(noise_cov.data.shape) == 1:
#             W_d = inv(noise_cov.data*np.identity(noise_cov.data.shape[0]))
#             # W_d = noise_cov.data*np.identity(noise_cov.data.shape[0])
#         else:
#             W_d = inv(noise_cov.data)
#             # W_d = noise_cov.data


#     # Calculate noise term using regularization parameter reg and the Data Space Metric
#     noise_term = (reg**2) * inv(W_d)

#     # noise-free case
#     # G = W_j_inv * leadfield.T * inv(leadfield * W_j_inv * leadfield.T) #* leadfield.T * W_d

#     # Calculate the inverse operator G
#     G = W_j_inv * leadfield.T * inv(leadfield * W_j_inv * leadfield.T + noise_term) #* leadfield.T * W_d

#     # calculate source
#     y_hat = np.array(np.matmul(G, x))
#     stc = mne.SourceEstimate(y_hat, vertices, tmin=evoked.tmin, 
#         tstep=1/evoked.info["sfreq"], subject=forward["src"]._subject,
#         verbose=verbose)
#     return stc

def compute_laura(evoked, forward, noise_cov=None, reg=200, drop_off=2,
    verbose=0):
    """ Calculate Local AUto-Regressive Averages (LAURA) inverse solution


    Parameters
    ----------
    evoked : instance of mne.Evoked
        The evoked data
    forward : instance of Forward
        The forward solution.
    noise_cov : instance of Covariance
        The noise covariance.
    reg : float
        Value that weights noise regularization. The higher, the more
        conservative the source localization will be. As a rule of thumb you can
        estimate the optimal reg parameter as follows: reg = 1e3 / SNR, whereas
        SNR corresponds to the signal-to-noise ratio in your data.
    drop_off : int/float
        Value that is often denoted as exponent "e_i" in the literature [2,
        Appendix] and is set to "2" at default. It corresponds to fields that
        decrease with the square of the inverse distance. Lowering this value
        (e.g., drop_off=1.5) will decrease sparsity of sources, whereas
        increasing this value (e.g., drop_off=2.5) will increase sparsity.
    verbose : bool,
        Controls verbosity.

    Returns
    -------
    stc : instance of SourceEstimate
        The source estimates.
    
    References
    ---------

    [1] Menendez, R. G. D. P., Andino, S. G., Lantz, G., Michel, C. M., &
    Landis, T. (2001). Noninvasive localization of electromagnetic epileptic
    activity. I. Method descriptions and simulations. Brain topography, 14(2),
    131-137.

    [2] de Peralta Menendez, R. G., Murray, M. M., Michel, C. M., Martuzzi, R.,
    & Andino, S. L. G. (2004). Electrical neuroimaging based on biophysical
    constraints. Neuroimage, 21(2), 527-539.

    
    """
    G = make_laura(forward, noise_cov=noise_cov, reg=reg, 
        drop_off=drop_off, verbose=verbose)
    stc = apply_laura(evoked, G, forward, verbose=0)
    return stc

def make_laura(forward, noise_cov=None, reg=200, drop_off=2,
    verbose=0):
    """ Calculate the Local AUto-Regressive Averages (LAURA) inverse operator


    Parameters
    ----------
    forward : instance of Forward
        The forward solution.
    noise_cov : instance of Covariance
        The noise covariance.
    reg : float
        Value that weights noise regularization. The higher, the more
        conservative the source localization will be. As a rule of thumb you can
        estimate the optimal reg parameter as follows: reg = 1e3 / SNR, whereas
        SNR corresponds to the signal-to-noise ratio in your data.
    drop_off : int/float
        Value that is often denoted as exponent "e_i" in the literature [2,
        Appendix] and is set to "2" at default. It corresponds to fields that
        decrease with the square of the inverse distance. Lowering this value
        (e.g., drop_off=1.5) will decrease sparsity of sources, whereas
        increasing this value (e.g., drop_off=2.5) will increase sparsity.
    verbose : bool,
        Controls verbosity.

    Returns
    -------
    stc : instance of SourceEstimate
        The source estimates.
    
    References
    ---------

    [1] Menendez, R. G. D. P., Andino, S. G., Lantz, G., Michel, C. M., &
    Landis, T. (2001). Noninvasive localization of electromagnetic epileptic
    activity. I. Method descriptions and simulations. Brain topography, 14(2),
    131-137.

    [2] de Peralta Menendez, R. G., Murray, M. M., Michel, C. M., Martuzzi, R.,
    & Andino, S. L. G. (2004). Electrical neuroimaging based on biophysical
    constraints. Neuroimage, 21(2), 527-539.

    
    """
    assert forward['surf_ori'], "dipole orientations in mne.Forward must be fixed!"
    forward = forward.copy()

    # make sure to select all the right channels
    if noise_cov is not None:
        noise_cov = noise_cov.copy()
        ch_names = list(set(noise_cov.ch_names).intersection(forward.ch_names))
        noise_cov = noise_cov.pick_channels(ch_names)
        forward = forward.pick_channels(ch_names)
    
    # Extract the necessary variables from the forward model
    pos = pos_from_forward(forward, verbose=verbose)
    leadfield = forward['sol']['data']
    
    
    # pairwise distance matrix of all vertex/dipole locations
    d = cdist(pos, pos)
    # Get the adjacency matrix of the source spaces
    adj = mne.spatial_src_adjacency(forward["src"], verbose=verbose).toarray()
    # set non-neighboring dipoles to zero
    for i in range(d.shape[0]):
        # find dipoles that are no neighbor to dipole i
        non_neighbors = np.where(~adj.astype(bool)[i, :])[0]
        # append dipole itself
        non_neighbors = np.append(non_neighbors, i)
        # set non-neighbors to zero
        d[i, non_neighbors] = 0
        # neighbors = np.where(adj.astype(bool)[i, :])[0]
        # d[i, neighbors] = -d[i, neighbors]**(drop_off)

    A = -d**-drop_off
    A[np.isinf(A)] = 0
    W = np.identity(A.shape[0])
    M = np.matmul(W, A)

    # Source Space metric
    W_j = np.matrix(inv(np.matmul(M.T, M)))
    W_j_inv = np.matrix(inv(W_j))

    # Data Space metric
    if noise_cov is None:
        W_d = np.matrix(np.identity(leadfield.shape[0]))
    else:
        # check if mne.Covariance is 1D (only diagonal values of the matrix)
        if len(noise_cov.data.shape) == 1:
            W_d = inv(noise_cov.data*np.identity(noise_cov.data.shape[0]))
            # W_d = noise_cov.data*np.identity(noise_cov.data.shape[0])
        else:
            W_d = inv(noise_cov.data)
            # W_d = noise_cov.data


    # Calculate noise term using regularization parameter reg and the Data Space Metric
    noise_term = (reg**2) * inv(W_d)

    # noise-free case
    # G = W_j_inv * leadfield.T * inv(leadfield * W_j_inv * leadfield.T) #* leadfield.T * W_d

    # Calculate the inverse operator G
    G = W_j_inv * leadfield.T * inv(leadfield * W_j_inv * leadfield.T + noise_term) #* leadfield.T * W_d
    return G

def apply_laura(evoked, G, forward, verbose=0):
    '''
    Parameters
    ----------
    evoked : instance of mne.Evoked
        The evoked data
    G : numpy.ndarray
        The generalized inverse operator. Can be computed with the funciton
        laura.make_laura
    forward : mne.Forward
        Instance of the mne.Forward object - forward model
    
    Return
    ------
    stc : mne.SourceEstimate
        The LAURA inverse solution as instance of mne.SourceEstimate
    '''
    # Extract necessary parameters
    evoked = evoked.copy().set_eeg_reference('average', projection=True).apply_proj()
    forward = forward.copy()
    vertices = [forward["src"][0]['vertno'], forward["src"][1]['vertno']]

    # calculate source
    x = evoked.data
    
    y_hat = np.array(np.matmul(G, x))
    stc = mne.SourceEstimate(y_hat, vertices, tmin=evoked.tmin, 
        tstep=1/evoked.info["sfreq"], subject=forward["src"]._subject,
        verbose=verbose)
    return stc